packages/devtools_app/lib/src/timeline/timeline_flame_chart.dart

```diff
@@ -333,6 +413,119 @@ class FullTimelineFlameChartCanvas extends FlameChartCanvas<FullTimelineData> {
         0;
     return absoluteY - topOffset - (section * sectionSpacing);
   }
+
+  @override
+  void paintCallback(CanvasRenderingContext2D canvas, Rect visible) {
+    paintSections(canvas, visible);
+    paintRows(canvas, visible);
+    _paintAsyncGuidelines(canvas, visible);
+    paintTimelineGrid(canvas, visible);
+  }
+
+  void _paintAsyncGuidelines(CanvasRenderingContext2D canvas, Rect visible) {
+    final firstVerticalGuideline = _binarySearchForVerticalLineSegment(visible);
+    final firstHorizontalGuideline =
+        _binarySearchForHorizontalLineSegment(visible);
+
+    // Only modify the canvas style if we have any guidelines to paint.
+    if (firstHorizontalGuideline != null || firstVerticalGuideline != null) {
+      canvas
+        ..strokeStyle = colorToCss(guidelineColor)
+        ..lineWidth = guidelineWidth;
+    }
+
+    if (firstVerticalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        verticalGuidelines,
+        firstVerticalGuideline,
+      );
+    }
+
+    if (firstHorizontalGuideline != null) {
+      _paintGuidelines(
+        canvas,
+        visible,
+        horizontalGuidelines,
+        firstHorizontalGuideline,
+      );
+    }
+  }
+
+  void _paintGuidelines(
+    CanvasRenderingContext2D canvas,
+    Rect visible,
+    List<LineSegment> guidelines,
+    LineSegment first,
+  ) {
+    final guidelinesToPaint = guidelines
+        .sublist(guidelines.indexOf(first))
+        .where((line) => line.intersects(visible));
+    for (var line in guidelinesToPaint) {
+      canvas
+        ..beginPath()
+        ..moveTo(line.start.x, line.start.y + topOffset)
+        ..lineTo(line.end.x, line.end.y + topOffset)
+        ..closePath()
+        ..stroke();
+    }
+  }
+
+  LineSegment _binarySearchForVerticalLineSegment(Rect rect) {
+    final LineSegmentSearchCondition shouldSearchFirstHalf =
+        (line, rect) => (line as VerticalLineSegment).x > rect.right;
+    final LineSegmentSearchCondition shouldSearchSecondHalf =
+        (line, rect) => (line as VerticalLineSegment).x < rect.left;
+    return _binarySearchForLineSegment(
+      verticalGuidelines,
+      rect,
+      shouldSearchFirstHalf,
+      shouldSearchSecondHalf,
+    );
+  }
+
+  LineSegment _binarySearchForHorizontalLineSegment(Rect rect) {
+    final LineSegmentSearchCondition shouldSearchFirstHalf =
+        (line, rect) => (line as HorizontalLineSegment).y > rect.bottom;
+    final LineSegmentSearchCondition shouldSearchSecondHalf =
+        (line, rect) => (line as HorizontalLineSegment).y < rect.top;
+    return _binarySearchForLineSegment(
+      horizontalGuidelines,
+      rect,
+      shouldSearchFirstHalf,
+      shouldSearchSecondHalf,
+    );
+  }
+
+  LineSegment _binarySearchForLineSegment(
```

Comment:
there should be an existing binary search algorithm we can use rather than reimplementing.

Reply:
https://github.com/dart-lang/collection probably has a fine implementation.

Reply:
yeah. You will want to create a synthetic LineSegment that indicates the leftmost line segment that can be in view.

Reply:
done